server/session/handler_item_stack_request.go

```diff
@@ -485,3 +556,184 @@ func (h *ItemStackRequestHandler) reject(id int32, s *Session) {
+		if err != nil {
+			return false
+		}
+		if oldSt.Empty() {
+			// We should still up the satisfied inputs count if both stacks are empty.
+			if inputs[satisfiedInputs].Empty() {
+				satisfiedInputs++
+			}
+
+			continue
+		}
+
+		// Items that apply to all types, so we just compare with the name and count.
+		currentInputToMatch := inputs[satisfiedInputs]
+		if currentInputToMatch.Variants {
+			name, _ := oldSt.Item().EncodeItem()
+			otherName, _ := currentInputToMatch.Item().EncodeItem()
+			if name == otherName && oldSt.Count() >= currentInputToMatch.Count() {
+				satisfiedInputs++
+			}
+		} else {
+			if oldSt.Comparable(currentInputToMatch.Stack) {
+				satisfiedInputs++
+			}
+		}
+	}
+
+	return satisfiedInputs == len(inputs)
+}
+
+// removeInventoryInputs removes the inputs in the player inventory.
+func (h *ItemStackRequestHandler) removeInventoryInputs(inputs []recipe.InputItem, s *Session) error {
+	inputMap := h.inputMapFromInputs(inputs)
+
+	updateStack := func(container byte, slot byte, oldSt item.Stack) {
+		if oldSt.Empty() {
+			return
+		}
+
+		name, meta := oldSt.Item().EncodeItem()
+		if input, ok := inputMap[name]; ok {
+			if input.Empty() {
+				return
+			}
+			_, otherMeta := input.Item().EncodeItem()
+			if meta == otherMeta || input.Variants {
+				if !input.Empty() {
+					targetRemoval := oldSt.Count()
+					if input.Count() < oldSt.Count() {
+						targetRemoval = input.Count()
+					}
+
+					st := oldSt.Grow(-targetRemoval)
+					h.setItemInSlot(protocol.StackRequestSlotInfo{
+						ContainerID:    container,
+						Slot:           slot,
+						StackNetworkID: item_id(st),
+					}, st, s)
+
+					input.Stack = input.Grow(-targetRemoval)
+					inputMap[name] = input
+				}
+			}
+		}
+	}
+
+	for slot, oldSt := range s.inv.Items() {
+		updateStack(containerFullInventory, byte(slot), oldSt)
+	}
+
+	offset := s.craftingOffset()
+	for i := byte(0); i < s.craftingSize(); i++ {
+		slot := i + offset
+
+		oldSt, err := s.ui.Item(int(slot))
+		if err != nil {
+			return err
+		}
+
+		updateStack(containerCraftingGrid, slot, oldSt)
+	}
+
+	return nil
+}
+
+// removeGridInputs removes the inputs passed in the crafting grid.
+func (h *ItemStackRequestHandler) removeGridInputs(inputs []recipe.InputItem, s *Session) error {
+	offset := s.craftingOffset()
+
+	var index int
+	for i := byte(0); i < s.craftingSize(); i++ {
+		if index == len(inputs) {
+			break
+		}
+
+		slot := i + offset
+		oldSt, err := s.ui.Item(int(slot))
+		if err != nil {
+			return fmt.Errorf("expected item doesn't exist: " + err.Error())
+		}
+		if oldSt.Empty() {
+			// We should still up the index if the expected input is empty.
+			if inputs[index].Empty() {
+				index++
+			}
+
+			continue
+		}
+
+		st := oldSt.Grow(-inputs[index].Count())
+		h.setItemInSlot(protocol.StackRequestSlotInfo{
+			ContainerID:    containerCraftingGrid,
+			Slot:           slot,
+			StackNetworkID: item_id(st),
+		}, st, s)
+		index++
+	}
+
+	return nil
```

Comment:
Lots of semi-duplicated code here. Do you reckon there's any way we could re-use some of the code so these functions can become smaller and easier to understand?

Reply:
I don't have any immediate ideas, but I might (probably) be able to find some ways to merge some bits of code together. Will try this later.